src/fsharp/PrettyNaming.fs

```diff
@@ -109,66 +109,158 @@ module internal Microsoft.FSharp.Compiler.PrettyNaming
+            | None ->
+                let opLength = op.Length
+                let sb = new System.Text.StringBuilder (opNamePrefix, opNamePrefix.Length + (opLength * maxOperatorNameLength))
+                for i = 0 to opLength - 1 do
+                    let c = op.[i]
+                    match t2.TryGetValue c with
+                    | true, x ->
+                        sb.Append(x) |> ignore
+                    | false, _ ->
+                        sb.Append(c) |> ignore
+
+                /// The compiled (mangled) operator name.
+                let opName = sb.ToString ()
+
+                // Cache the compiled name so it can be reused.
+                lock compiledOperators <| fun () ->
+                    compiledOperators.[op] <- opName
+
+                opName
+
+    /// Compiles a custom operator into a mangled operator name.
+    /// For example, "!%" becomes "op_DereferencePercent"
     let CompileOpName =
-        let t = Map.ofList opNameTable
-        let t2 = Map.ofList opCharTranslateTable
-        fun n -> 
-            match t.TryFind(n) with 
-            | Some(x) -> x 
-            | None -> 
-                if IsOpName n then 
-                  let mutable r = []
-                  for i = 0 to String.length n - 1 do
-                      let c = n.[i]
-                      let c2 = match t2.TryFind(c) with Some(x) -> x | None -> string c
-                      r <- c2 :: r 
-                  "op_"^(String.concat "" (List.rev r))
-                else n
-
-    let IsMangledOpName (n:string) = n.Length >= 3 && n.Substring(0,3) = "op_"     
-                             
-    let DecompileOpName = 
-      let t = new Dictionary<string,string>()
-      for (x,y) in opNameTable do
-          t.Add(y,x)
-      fun n -> 
-          let mutable res = Unchecked.defaultof<_>
-          if t.TryGetValue(n,&res) then 
-              res
-          else
-              if n.StartsWith("op_",System.StringComparison.Ordinal) then 
-                let rec loop (remaining:string) = 
+        /// Maps the built-in F# operators to their mangled operator names.
+        let standardOpNames =
+            let opNames = Dictionary<_,_> (opNameTable.Length, System.StringComparer.Ordinal)
+            for x, y in opNameTable do
+                opNames.Add (x, y)
+            opNames
+
+        fun op ->
+            match standardOpNames.TryGetValue op with
+            | true, x -> x
+            | false, _ ->
+                if not <| IsOpName op then op
+                else
+                    compileCustomOpName op
+
+    /// Decompiles the mangled name of a custom operator back into an operator.
+    /// For example, "op_DereferencePercent" becomes "!%"
+    let private decompileCustomOpName =
+        // Memoize this operation. Custom operators are typically used more than once
+        // so this avoids repeating decompilation.
+        let decompiledOperators = Dictionary<_,_> (System.StringComparer.Ordinal)
```

Comment:
Any reason not to use ConcurentDictionary?  I'd prefer that unless there's a proven case against it.